src/data/list/rotate.lean

```diff
@@ -376,6 +395,120 @@ begin
   exact ⟨λ ⟨n, hn, h⟩, ⟨n, nat.lt_succ_of_le hn, h⟩, λ ⟨n, hn, h⟩, ⟨n, nat.le_of_lt_succ hn, h⟩⟩
 end
 
+/-- List of all cyclic permutations of `l`.
+The `cyclic_permutations` of a nonempty list `l` will always contain `length l` elements.
+This implies that under certain conditions, there are duplicates in `cyclic_permutations l`.
+The `n`th entry is equal to `l.rotate (n + 1)`, proven in `nth_le_cyclic_permutations`.
+The proof that every cyclic permutant of `l` is in the list is `mem_cyclic_permutations_iff`.
+
+     cyclic_permutations [1, 2, 3, 2, 4] =
+       [[1, 2, 3, 2, 4], [2, 3, 2, 4, 1], [3, 2, 4, 1, 2],
+        [2, 4, 1, 2, 3], [4, 1, 2, 3, 2]] -/
+def cyclic_permutations : list α → list (list α)
+| []       := [[]]
+| (x :: l) := tail (zip_with (++) (tails (x :: l)) (inits (x :: l)))
+
+@[simp] lemma cyclic_permutations_nil : cyclic_permutations ([] : list α) = [[]] := rfl
+
+@[simp] lemma cyclic_permutations_cons (x : α) (l : list α) :
+  cyclic_permutations (x :: l) = tail (zip_with (++) (tails (x :: l)) (inits (x :: l))) := rfl
+
+lemma length_cyclic_permutations_cons (x : α) (l : list α) :
+  length (cyclic_permutations (x :: l)) = length l + 1 :=
+by simp
+
+@[simp] lemma nth_le_cyclic_permutations (l : list α) (n : ℕ)
+  (hn : n < length (cyclic_permutations l)) :
+  nth_le (cyclic_permutations l) n hn = l.rotate (n + 1) :=
```

Comment:
It seems to me that
```suggestion
  nth_le (cyclic_permutations l) n hn = l.rotate n :=
```
Would me a nicer property. Does this make things inconvenient in a way that I'm missing?